rust/kernel/cpu.rs

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! CPU utilities for SMP systems.
+//!
+//! C header: [`include/linux/smp.h`](../../../../include/linux/smp.h).
+//!
+//! Note: This module is a work in progress.
+
+use crate::{bindings, error::code::*, prelude::*, to_result};
+use core::{marker::PhantomData, ops::Deref};
+
+/// Represents an SMP CPU by its zero-based kernel processor id.
+///
+/// Cpu values can be safely passed between threads (and across physical CPUs) as they are
+/// merely an integer index representing a CPU. All actual interaction with CPUs is late-
+/// bound.
+#[derive(Copy, Clone)]
+pub struct Cpu(u32);
+
+struct FnCall<F, R>
+where
+    F: FnOnce() -> R,
+{
+    out: Option<R>,
+    func: Option<F>,
+}
+
+unsafe extern "C" fn call_fn<F, R>(fn_box: *mut core::ffi::c_void)
+where
+    F: FnOnce() -> R,
+{
+    // SAFETY: The casting here is the inverse of the cast to *mut c_void that occurs in Cpu::call
+    // below. That call waits for this call to complete ensuring fn_call is live for the duration.
+    let fn_call = unsafe { &mut *(fn_box as *mut FnCall<F, R>) };
+    let func = fn_call.func.take().unwrap();
+    fn_call.out = Some(func());
+}
+
+impl Cpu {
+    /// Constructs a new instance referring to a specific processor by id.
+    ///
+    /// This function checks that the passed id is a valid _potential_ processor id, but does
```

Comment:
Since the first "id" in this line refers to the parameter, while the second "id" refers to the concept itself, it may be more clear (in particular when rendered) like this:

```suggestion
    /// This function checks that the passed `id` is a valid _potential_ processor id, but does
```